hilbert transform

joglekara opened the pull request:
Follows SciPy's implementation except for the `ndim > 1` part.

Tested locally to reproduce `osp.signal.hilbert` for even and odd N but could use some guidance on how to implement a proper test

hawkinsp committed:
Increase minimum NumPy version to 1.21.

Also increase minimum SciPy version to 1.7, which was released just before NumPy 1.21.

b0nce committed:
Fix loc and scale parameters in scipy.logistic. Add CDF and SF for several distributions.

froystig committed:
`custom_vjp` symbolic zeros support

mattjj committed:
make mlir arg and result names work with pmap

This is a follow-up on #15080 to restore (and indeed fix!) how pmap builds a
jaxpr with debug info (i.e. parameter names and result paths). The difference
with the machinery in #15080 is just to deal with pmap being final-style (i.e.
build the jaxpr at the last second, well after pytrees have been flattened away
and transformations have been applied), whereas the machinery for pjit in
imagine, plumbing for the former is a bit more long-range and subtle.

The main idea here is that we need to annotate and maintain debug info on the
lu.WrappedFun instance, which we first form at the api.py level, then pass
through transformations (which can either update or drop debug info), then
finally hand off to the impl rule to be traced to a jaxpr. It makes sense as an
annotation, parallel with the in_type annotation used for dynamic shapes,
because the debug info has to be updated as transformations are applied, since
they might e.g. add tangent inputs and outputs.

In more detail: with an initial-style higher-orer primitive (like pjit), a
jaxpr is formed immediately. Transformations, like autodiff, are
jaxpr-to-jaxpr, and so those transformations (like ad.jvp_jaxpr) need to return
a new jaxpr either with updated debug info or no debug info at all. (The initial
implementation in #15080 doesn't provide updated debug info in any of those
jaxpr-to-jaxpr transformation functions, so the debug info is only applied to
the jaxpr and then lowered to MLIR when the pjit as at the top level.)

For final-style, like pmap here, instead of transformations being
jaxpr-to-jaxpr, they're WrappedFun-to-WrappedFun. And so, analogously,
transformations, like ad.JVPTrace.process_map, would need to produce a
WrappedFun with updated debug info or no debug info at all. (ALso analogously
to #15080, this PR only implements enough for the debug info to be preserved
for top-level pmaps.)

This PR doens't yet delete the trace-time debug info in partial_eval.py. But
that'll happen too!

JiaYaobo committed:
add wald random generator

add wald to random.py

gnecula committed:
[jax2tf] Fix grad of pjit in native lowering.

Since jax2tf.convert is called recursively for the purpose of
serializing the vjp function, we must ensure that if the primal
function is a pjit with shardings then the vjp function must also
be converted as a pjit.

Without this fix the serialization with gradients of a pjit function
will fail the an error that there are shardings but not pjit at
the top-level.

nouiz committed:
Fix inspect_array_sharding with grad.

joglekara committed:
hilbert transform

joglekara committed:
newline

yashk2810 committed:
Raise a better error message when there is a device assignment mismatch via the apply_primitive route.

PiperOrigin-RevId: 518282464

jakevdp committed:
jnp.mean: fix incorrect return value for large arrays

hawkinsp committed:
Relax type annotations on lax slicing functions.

PiperOrigin-RevId: 518308356

jakevdp commented:
Thanks! I think as far as testing goes, you could follow some of the FFT tests as a template: https://github.com/google/jax/blob/4bd7f408c60266641ea8db2957ea458edcdcf50d/tests/fft_test.py#L231

But the tests for functions in this file should go in https://github.com/google/jax/blob/main/tests/scipy_signal_test.py

gnecula committed:
[jax2tf] Clean up the jax2tf sharding_tests.

The sharding tests are very important because jax2tf must do non-trivial manipulation of shardings, e.g., to wrap the inputs and the outputs with
sharding annotations.

Merged two test classes ShardedJitHloTest and Sharding. One of the classes was
just checking the annotations in the TF HLO, the other one was just running
the code and comparing the results. Now we do both in one test.

Refactored the code to log JAX HLO and to check the occurrences of annotations
in TF HLO. Now we support checking that the occurrence count is equal to a value
or greater or equal to a value.

Added more annotation checking. This makes the test more informative (because
there is no other good way to check that sharding was applied correctly).
But this makes it also possible that the test will fail when we change the
JAX lowering.

PiperOrigin-RevId: 518362978

yashk2810 committed:
Remove the config.jax_array and jax_jit_pjit_api_merge flag usage since those are always True

PiperOrigin-RevId: 518368963

jakevdp committed:
Document ShapeDtypeStruct

atondwal committed:
[jax2tf] Add back_compat test for LuDecomposition

PiperOrigin-RevId: 518412422

atondwal committed:
LuDecomposition moved from fallback path to custom_call

PiperOrigin-RevId: 518417901

Commit:
[PJRT C API] Add parsing PJRT client create options from json file.

PiperOrigin-RevId: 518418760

mattjj committed:
enable pjit recursive typechecking

Give pjit_p a custom typecheck rule, which basically just calls the
core._check_call utility (which was made for xla_call_p and core.call_p).

This revealed the need for a slight generalization of the custom_typecheck rule
signature, for better "context-aware" printing of jaxpr type errors: the rules
should have a `ctx_factory` first argument. **The reason this PR touches so
many files is just that it makes the trivial tweaks to all existing typecheck
rules to accomodate that new signature.** I didn't adapt any other higher-order
primitives' rules to actually use the context, but presumably errors for HOPs
like scan would be improved by using it. Follow-up work!

It's key that core._check_call works with dynamic shapes; this PR is soon to be
followed by some djax+pjit PRs!

hawkinsp committed:
Fix mypy failures in jax2tf.

PiperOrigin-RevId: 518572905

hawkinsp committed:
Revert: `custom_vjp` symbolic zeros support

PiperOrigin-RevId: 518597609

jakevdp committed:
jax.random: remove scale from wald function

jakevdp committed:
DOC: add formulae for distributions in jax.random

GleasonK committed:
Improve handling of dynamic shapes in jax native serialization

PiperOrigin-RevId: 518634912

nouiz committed:
WAR the dependency issue in the nightly CI container.

mattjj committed:
add experimental jax_log_checkpoint_residuals option

The main idea here is to improve tooling for knowing what residuals are being
saved and why. There's a lot more that can be done here (e.g. naming the
arguments, explaining what JVP rule produced these residuals, explaining what
consumed them, etc) but this is a start.

Co-authored-by: Qiao Zhang <zhangqiaorjc@google.com>

sharadmv committed:
Add print statement to help debug spurious test failure

pschuh committed:
Redefine `compile_and_serialize` as `serialize(lowered.compile())`.

This has the downside of keeping around the UnloadedMeshComputation,
but it makes the serialize() API easier to understand.

PiperOrigin-RevId: 518715469

mattjj committed:
[dynamic-shapes] make dynamic shape staging-to-jaxpr work with pjit

Commit:
fix typo: "one of more" -> "one or more"

PiperOrigin-RevId: 518762341

Commit:
Fix isinstance(k, PRNGKeyArray) on PRNGKeyArray subclasses

PiperOrigin-RevId: 518803946

gnecula committed:
[jax2tf] Fix tests broken by upgrade of XlaCallModule

PiperOrigin-RevId: 518811580

hawkinsp committed:
[XLA:Python] Change JAX and the XLA Python extension to get NumPy bfloat16/float8 types from ml_dtypes.

PiperOrigin-RevId: 518830467

hawkinsp committed:
[XLA:Python] Allow passing ExecutableBuildOptions to outfeed receiver.

Outfeed receiver compiles computations (during shutdown), and if the correct options aren't provided, then it may not be able to do things like find ptxas for CUDA builds. Plumb the executable build options through from Python.

PiperOrigin-RevId: 518852909

nouiz committed:
Add missing file

hawkinsp committed:
Move jax._src.typing into a separate Bazel target.

PiperOrigin-RevId: 518899136

jakevdp committed:
internal: refactor array methods into separate private submodule

yashk2810 committed:
Deprecated xla_call_p since it has been replaced with pjit.pjit_p

PiperOrigin-RevId: 518921538

nouiz committed:
WAR ssh timeout like:
client_loop: send disconnect: Broken pipe
https://github.com/google/jax/actions/runs/4500333187/jobs/7919324156#step:8:42

jakevdp committed:
Fix mypy issue in jax/experimental/jet.py

cgarciae committed:
add trailing-whitespace pre-commit hook

mattjj committed:
improve scan error messages

gnecula committed:
[jax2tf] Minor addition to the documentation

PiperOrigin-RevId: 518969936

atondwal committed:
Refactor special functions into their own module.

We're going to want to decompose these using series and
continued fraction representations, and for that we'll need
control flow

PiperOrigin-RevId: 518977008

gnecula committed:
[jax2tf] Create a jax_export library with JAX-only pieces for native serialization

This is a pure refactor, no functionality should change.

PiperOrigin-RevId: 518982222

maxwillzq committed:
Add padding option "SAME_LOWER" for ticket https://github.com/google/jax/pull/14990

PiperOrigin-RevId: 518984018

gnecula committed:
[jax2tf] A simple failing test on TPU with native serialization

PiperOrigin-RevId: 518987577

atondwal committed:
Migrate igamma_p off xla_fallback

We decompose it into a series or a call to igammac.

PiperOrigin-RevId: 518993077

jakevdp committed:
lax_numpy: move quantile-based functions to reductions.py

skye committed:
Remove PJRT C API bypass.

Now that all functionality needed by frameworks is implemented, let's
remove the possibility of not noticing missing functionality due to
the bypass.

PiperOrigin-RevId: 519018438

mattjj committed:
fix jax.Array.round()

fixes #15190

atondwal committed:
[jaxlib] fix build w/ depenency on stablehlo_serialization

PiperOrigin-RevId: 519120624

atondwal committed:
Migrate igammac_p off xla_fallback path

It is now decomposed into stablehlo ops.

PiperOrigin-RevId: 519122775

nouiz committed:
remove another dependency not currently needed.

atondwal committed:
Migrate igamma_grad_a_p off xla_fallback

PiperOrigin-RevId: 519148548

atondwal committed:
Migrate random_gamma_grad off xla_fallback

PiperOrigin-RevId: 519154537

joglekara committed:
dtype errors

joglekara commented:
I wrote some of those tests :) Looks like ya'll have improved on them substantially!

```python
E       jax._src.dtypes.TypePromotionError: Input dtypes ('float32', 'complex64') have no available implicit dtype promotion path when jax_numpy_dtype_promotion=strict. Try explicitly casting inputs to the desired output type, or set jax_numpy_dtype_promotion=standard.
```

I'm getting some of these though and I can't quite grok how to make sure both sides are the same dtype. Can I get an assist please?

yashk2810 committed:
After the SPMD bug fix, always take the _rewriting_take route for getitem instead of bouncing to host.

PiperOrigin-RevId: 519170785

jakevdp commented on jax/_src/scipy/signal.py:
```diff
@@ -637,3 +637,31 @@ def istft(Zxx: Array, fs: ArrayLike = 1.0, window: str = 'hann',
+    # h[0] = 1
+    # h[1:(N + 1) // 2] = 2
+
+  if x.ndim > 1:
```
I would move this to the top of the function, so the error is raised before doing all the computations on `x`.

joglekara replied:
done

jakevdp commented on jax/_src/scipy/signal.py:
```diff
@@ -637,3 +637,31 @@ def istft(Zxx: Array, fs: ArrayLike = 1.0, window: str = 'hann',
+
+  Xf = jnp.fft.fft(x, N, axis=axis)
+  if N % 2 == 0:
+    h = jnp.concatenate([jnp.ones(1), jnp.ones(N // 2 - 2) * 2, jnp.ones(1), jnp.zeros(N // 2)])
```
For type safety / efficiency, I would explicitly construct these arrays with `Xf.dtype`. And use `jnp.full(..., 2)` rather than `jnp.ones(...) * 2`

Also, is the second term correct? If `N < 4` than this is attempting to construct an array with negative length.

Does the algorithm require that `N >= 4`? If so we should check that and raise an appropriate error.

jakevdp replied:
Or maybe simpler to write something like this:
```python
h = jnp.zeros(N).at[0].set(1).at[1:N // 2].set(2).at[N // 2].set(1)
```

jakevdp commented on jax/_src/scipy/signal.py:
```diff
@@ -637,3 +637,31 @@ def istft(Zxx: Array, fs: ArrayLike = 1.0, window: str = 'hann',
+    # h[0] = h[N // 2] = 1
+    # h[1:N // 2] = 2
+  else:
+    h = jnp.concatenate([jnp.ones(1), jnp.ones((N + 1) // 2 - 1) * 2, jnp.zeros(N // 2)])
```
Same broad comments here (type safety, negative array lengths. Also, I think this could probably be done without an if statement. The second array length would be `N - 1 - N // 2` in either case, yes?

joglekara replied:
hmm, I'm mostly just transcribing the `osp` implementation. It is working now, and I think it does need the `if` statement to address the last point given the suggested rewrite

jakevdp commented on tests/scipy_signal_test.py:
```diff
@@ -390,5 +391,24 @@ def testIstftAgainstNumpy(self, *, shape, dtype, fs, window, nperseg,
+    jnp_op = getattr(jsp_signal, name)
+    osp_op = getattr(osp_signal, name)
+    jnp_fn = lambda a: jnp_op(a, N=N, axis=axis)
+    osp_fn = lambda a: osp_op(a, N=N, axis=axis)
```
Since `name` is not changing, it would be simpler to write:
```python
    jnp_fn = lambda a: jsp_signal.hilbert(a, N=N, axis=axis)
    osp_fn = lambda a: osp_signal.hilbert(a, N=N, axis=axis)
```

joglekara replied:
indeed!

jakevdp commented on tests/scipy_signal_test.py:
```diff
@@ -390,5 +391,24 @@ def testIstftAgainstNumpy(self, *, shape, dtype, fs, window, nperseg,
+    # Numpy promotes to complex128 aggressively.
+    self._CheckAgainstNumpy(osp_fn, jnp_fn, args_maker, check_dtypes=False,
+                            tol=1e-4)
+    self._CompileAndCheck(jnp_op, args_maker)
```
I think you meant `jnp_fn` here

joglekara replied:
yup, thanks :)

jakevdp commented on jax/_src/scipy/signal.py:
```diff
@@ -637,3 +637,31 @@ def istft(Zxx: Array, fs: ArrayLike = 1.0, window: str = 'hann',
+
+@_wraps(osp_signal.hilbert)
+def hilbert(x, N=None, axis=-1):
+  x = jnp.asarray(x)
```
Please call `check_arraylike` before converting to `jnp.array`

joglekara replied:
done

jakevdp commented on jax/_src/scipy/signal.py:
```diff
@@ -637,3 +637,31 @@ def istft(Zxx: Array, fs: ArrayLike = 1.0, window: str = 'hann',
   return time, x
+
+@_wraps(osp_signal.hilbert)
+def hilbert(x, N=None, axis=-1):
```
As a public function, this should have type annotations

joglekara replied:
added

jakevdp reviewed (changes requested):
Looks great! A few comments below. Also please add `hilbert` here so it will appear in the HTML docs: https://github.com/google/jax/blob/main/docs/jax.scipy.rst#jaxscipysignal

pschuh committed:
Guard ArrayImpl checks by xla_extension_version.

PiperOrigin-RevId: 519191714

maxwillzq committed:
[1/n] store embedded tf.graph to stablehlo.custom_call

PiperOrigin-RevId: 519194911

hawkinsp committed:
Delete remote TPU support.

TPU VMs are the only supported way to use TPUs as of JAX 0.4.0.

PiperOrigin-RevId: 519211267

joglekara committed:
passing tests

jakevdp commented on jax/_src/scipy/signal.py:
```diff
@@ -637,3 +637,33 @@ def istft(Zxx: Array, fs: ArrayLike = 1.0, window: str = 'hann',
   return time, x
+
+@_wraps(osp_signal.hilbert)
+def hilbert(x: Array, N: int = None, axis: int = -1):
```
`N` needs to be `Optional[int]`

joglekara replied:
I like `Optional[int]` too, lets just go with that

jakevdp replied:
`Optional[int]` is more idiomatic

joglekara committed:
linters

joglekara committed:
Optional int

joglekara committed:
docs

joglekara committed:
Merge branch 'google:main' into hilbert
